### DISCUSSION TITLE
SCons Not Carrying Out Commands

### DISCUSSION THREAD

**Discussion by Jillinger** [SCons/scons#4465] [Q&A]
I'm not sure if this is a Godot related problem, but I gave SCons commands and it ignores them.
I used [Godot build options generator](https://godot-build-options-generator.github.io/) to generate an SCons custom option file.
I gave SCons the location of the file:
`scons profile="C:/develop/custom.py"`
SCons doesn't carry out the command. So I had to give the full list of commands in command prompt.

`scons platform=windows dev_build=yes arch=x86_64 vsproj=yes tools=yes mono_glue=no production=yes disable_advanced_gui=no deprecated=no minizip=no module_basis_universal_enabled=no module_camera_enabled=no module_enet_enabled=no module_jpg_enabled=no module_mbedtls_enabled=no module_meshoptimizer_enabled=no module_mobile_vr_enabled=no module_mono_enabled=yes module_msdfgen_enabled=no module_multiplayer_enabled=no module_noise_enabled=no module_openxr_enabled=no module_opus_enabled=no module_pvr_enabled=no module_squish_enabled=no module_text_server_adv_enabled=no module_theora_enabled=no module_upnp_enabled=no module_vorbis_enabled=no module_webm_enabled=no module_webrtc_enabled=no module_websocket_enabled=no module_webxr_enabled=no custom_modules="C:/develop/cppModules" mono_prefix="C:/Program Files/Mono"`

I got an error on my custom modules, so I removed the command `custom_modules="C:/develop/cppModules"`
However, SCons gives me the same error, though it shouldn't be reading from that directory anymore.
Is it using a cache? I tried using `--no-cache` , but that makes no difference.

**Comment by mwichmann** [comment]
For the first part, not handling your `profile=` argument, the Godot logic to handle that is pretty simple, I see. It's in `SConstruct`:
```py
profile = ARGUMENTS.get("profile", "")
if profile:
    if os.path.isfile(profile):
        customs.append(profile)
    elif os.path.isfile(profile + ".py"):
        customs.append(profile + ".py")

opts = Variables(customs, ARGUMENTS)
```
you might add some debug prints there to see if it actually got added, or if not, why.


To the second part, the persistence of something you thought you removed...  SCons does cache various things in different ways, but none that should apply here. The thing affected by `--no-cache` is the derived-file cache - that is, targets that successfully got built. That's off by default anyway (don't know if Godot changes that), and it doesn't seem that you're failing due to built things.  Though you didn't share any error messages...

Probably the Godot community would be best positioned to help you on this. We here are pretty much stuck in the absence of knowledge of what "I got an error" is

It's *possible* the `vsproj=yes` could be implicated in why you're seeing something you thought is disabled, if it wrote a Visual Studio project file first time and is somehow using it to drive subsequent builds. That seems pretty unlikely but if you see a project file in your directory and remove it, does that change the behavior?

**Reply by Jillinger** [reply]
> For the first part, not handling your profile= argument, the Godot logic to handle that is pretty simple, I see. It's in SConstruct:
> 
> profile = ARGUMENTS.get("profile", "")
> if profile:
>     if os.path.isfile(profile):
>         customs.append(profile)
>     elif os.path.isfile(profile + ".py"):
>         customs.append(profile + ".py")
> 
> opts = Variables(customs, ARGUMENTS)
> you might add some debug prints there to see if it actually got added, or if not, why.

I'm using command prompt, so I am not able to debug. Or can I. I don't know how.
I know for sure it wasn't added, because I had this in the custom.py file - `mono_prefix = "C:/Program Files/Mono"`.
However, SCons could not find mono, until I gave it this command - `mono_prefix="C:/Program Files/Mono"`

> you didn't share any error messages...

The errors:
```
C:/develop/cppModules/Summator/register_types.h(11): error C2065: 'ModuleInitializationLevel': undeclared identifier
C:/develop/cppModules/Summator/register_types.h(11): error C2146: syntax error: missing ')' before identifier 'p_level'
C:/develop/cppModules/Summator/register_types.h(12): error C2065: 'ModuleInitializationLevel': undeclared identifier
C:/develop/cppModules/Summator/register_types.h(12): error C2146: syntax error: missing ')' before identifier 'p_level'
modules\register_module_types.gen.cpp(197): error C3861: 'register_ModTest_types': identifier not found
modules\register_module_types.gen.cpp(200): error C3861: 'register_Summator_types': identifier not found
modules\register_module_types.gen.cpp(345): error C3861: 'unregister_ModTest_types': identifier not found
modules\register_module_types.gen.cpp(348): error C3861: 'unregister_Summator_types': identifier not found
[Initial build] scons: *** [modules\register_module_types.gen.windows.tools.x86_64.obj] Error 2←[94mCompiling ←[95m==> ←[93mmodules\bmp\image_loader_bmp.cpp←[0mimage_loader_bmp.cppscons: building terminated because of errors.
[Time elapsed: 00:01:40.493]
```

I used the Custom Modules example from the Godot Documentation - [Custom modules in C++](https://docs.godotengine.org/en/4.2/contributing/development/core_and_modules/custom_modules_in_cpp.html) and the one from GameFromScratch - [Creating a Custom Module in Godot](https://gamefromscratch.com/creating-a-custom-module-in-godot/).
Basically copied and pasted.


> Probably the Godot community would be best positioned to help you on this. We here are pretty much stuck in the absence of knowledge of what "I got an error" is

I thank you for your quick response. This is the reason I ask on Github.
If you go to Godot Community Forums, you can see my queries, and when I posted, and you can see what responses I got.
[Errors Using SCons to Create a VS Solution](https://forum.godotengine.org/t/errors-using-scons-to-create-a-vs-solution/43941/1)
[Help Compiling the GDNative Library](https://forum.godotengine.org/t/help-compiling-the-gdnative-library/44111/2)
[Trying to Compile Godot with SCons is not Easy](https://forum.godotengine.org/t/trying-to-compile-godot-with-scons-is-not-easy/44336/1)


> It's possible the vsproj=yes could be implicated in why you're seeing something you thought is disabled, if it wrote a Visual Studio project file first time and is somehow using it to drive subsequent builds. That seems pretty unlikely but if you see a project file in your directory and remove it, does that change the behavior?

Thanks. I'll check that.

**Comment by mwichmann** [comment]
> Thank you. I'm using the command prompt, so I don't think I can debug. Or can I.

You can do something like this - edit the `SConstruct` so that:

```py
profile = ARGUMENTS.get("profile", "")
if profile:
    if os.path.isfile(profile):
        customs.append(profile)
    elif os.path.isfile(profile + ".py"):
        customs.append(profile + ".py")
breakpoint()  # <- add this

opts = Variables(customs, ARGUMENTS)
```

It should stop in the debugger, at that point.  Print the value of `customs`. Here's a scenario where I point to a made-up (invalid) file:

```
$ scons profile=/tmp/foo
scons: Reading SConscript files ...
> /home/mats/github/godot/SConstruct(163)<module>()
-> opts = Variables(customs, ARGUMENTS)
(Pdb) p profile
'/tmp/foo'
(Pdb) p customs
['custom.py']
```

There it shows it did see the argument `profile=/tmp/foo`, but it didn't add it to `customs` because it didn't find the file, either with or without a `.py` suffix.

I guess I'm not sure what you do with that information, though... unless it tells you the path was somehow invalid, in which case at least there's a course of action for that problem.

**Reply by Jillinger** [reply]
I got.
```
(Pdb) p profile
''
(Pdb) p customs
['custom.py']
```

I guess you're right.
Unless you know how I can erase SCons memory, so it doesn't go into my custom modules directory.
I just had an idea. I'll just delete the Godot source, and start fresh.

**Reply by bdbaddog** [reply]
Really the issue you're having is not with SCons per say, but rather with how Godot is using it. Hopefully someone there with more knowledge of their implementation can give you a hand.

**Reply by Jillinger** [reply]
> Hopefully someone there with more knowledge of their implementation can give you a hand.

Hopefully. Thanks.
